# Executive Summary

This report describes the approach taken to build a gym subscription prediction model based on the user data provided as an input. It consist of descriptions, as well as the code written in Python language used to process data, build a model, and obtain final results. First section of this report serves to access data, get familiar with it, clean it, and transform it for the model to read it. Second section explains which model was selected and how it was build, compiled, and fit. Next section presents model quality assessment approach. Then, it is shown how the test file was scored with the prepared model. Last two sections describe findings encountered during the process and limitations of selected approach.

# Input Data and Transformations

In this section the process of analysis of the provided data is described. For this task I am going to store data in a format of so-called dataframes (DataFrame classes) from Pandas library wirtten for Python language:

In [206]:
import pandas as pd
#import json

### Access data

The data provided for this task has been uploaded in order to automatize this report. To access data I am going to clone the files that I stored in the git repository:

In [207]:
!git clone https://pkarczma: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@github.com/pkarczma/gym-subscription-predictor.git

fatal: destination path 'gym-subscription-predictor' already exists and is not an empty directory.


Next, one needs to read it. I am going to start with the data used for training. The data consist of CSV and JSON files, which are going to be read into two dataframes:

In [208]:
path = 'gym-subscription-predictor/'
df_csv = pd.read_csv(path+'train.csv')
df_json = pd.read_json(path+'train.json', orient='split').set_index('id')

### Merge data

The data for each used is split into two files: structured CSV file and unstructured JSON file. The data read from JSON file is unstructured. What we would like to achieve is one dataframe that consist of rows containing all the information about each user. To do that the data read from JSON file needs some conversion in order to extract necessary data nested inside. A new dataframe containing a list of group names for each user will be extracted. The information about the date of joining the group is skipped as this seem redundant for this task.

In [209]:
df_groups = pd.DataFrame(columns=['groups'])
for i in df_json.to_dict()['groups'].items():
  groups = ''
  for j in i[1]['data']:
    if len(groups) > 0:
      groups += '|'
    groups += j['group_name']
  df_groups = df_groups.append({'groups': groups}, ignore_index=True)

We can now merge two dataframes into one dataset containing all information about each user:

In [210]:
df = pd.concat([df_csv, df_groups], axis=1)

### Analyse and clear data

In this section we are going to get familiar with the data. We can have a look at some information about the dataframe and its columns, as well as look at the first few rows of the dataframe:

In [211]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   user_id                   4000 non-null   int64  
 1   target                    4000 non-null   int64  
 2   name                      3982 non-null   object 
 3   sex                       3616 non-null   object 
 4   dob                       3606 non-null   object 
 5   location                  4000 non-null   object 
 6   location_population       4000 non-null   int64  
 7   location_from             4000 non-null   object 
 8   location_from_population  4000 non-null   int64  
 9   occupation                4000 non-null   object 
 10  hobbies                   3320 non-null   object 
 11  daily_commute             3595 non-null   float64
 12  friends_number            4000 non-null   int64  
 13  relationship_status       3607 non-null   object 
 14  educatio

user_id  ...                                             groups
0        0  ...  Let's excercise together and lose a few kilo q...
1        1  ...  Tom Cruise group|Babysitters (Sokółka)|Work ab...
2        2  ...  Polish wildlife - best places|Politics and pol...
3        3  ...  The Aspiring Writer|Nutrition & food advices|G...
4        4  ...  The ultimate house and electro group|Pirates o...

[5 rows x 17 columns]

From this overview we can extract many useful information. We have 4000 users and 16 columns describing each user. There are some columns with empty cells, but we are going to deal with it later. At this point it would be useful to evaluate which part of the data isn't necessary for the model. I am going to assume that the following columns are not indicating an initial interest in gym subscription:
* name
* location_population
* location_from_population
* daily_commute
* credit_card_type

As a result I am going to drop them from the dataframe:

In [212]:
df = df.drop(columns=['name', 'location_population', 'location_from_population', 'daily_commute', 'credit_card_type'])

Afterwards, we can focus on the missing values that are still in the remaining data. We can count the number of missing values in each of the remaining columns:

In [213]:
df.isnull().sum(axis = 0)

user_id                  0
target                   0
sex                    384
dob                    394
location                 0
location_from            0
occupation               0
hobbies                680
friends_number           0
relationship_status    393
education              408
groups                   0
dtype: int64

There are still some missing values in data in several columns. I am going to use a different approach depending on the column with the missing values. The following procedure will be applied:
* 'user_id' / 'target' / 'location' / 'location_from' / 'occupation' / 'friends_number': no missing values, columns are useful, nothing changes
* 'education': fill missing falues with a median of a column
* 'hobbies': I assume that empty value means that a user has no hobbies, so I will fill missing values with empty string

For the remaining data with missing values it is problematic to deduce how replace it. One can also notice that the fraction of missing values to all values isn't too large - it is around 10% for each of the remaining columns with missing values. Consequently, the rows with at least one missing calue will be dropped from the dataset.



In [214]:
df['hobbies'] = df['hobbies'].fillna('')
df['education'] = df['education'].fillna(df['education'].median())
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2945 entries, 1 to 3999
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              2945 non-null   int64  
 1   target               2945 non-null   int64  
 2   sex                  2945 non-null   object 
 3   dob                  2945 non-null   object 
 4   location             2945 non-null   object 
 5   location_from        2945 non-null   object 
 6   occupation           2945 non-null   object 
 7   hobbies              2945 non-null   object 
 8   friends_number       2945 non-null   int64  
 9   relationship_status  2945 non-null   object 
 10  education            2945 non-null   float64
 11  groups               2945 non-null   object 
dtypes: float64(1), int64(3), object(8)
memory usage: 299.1+ KB


As a result, we removed around 25% of all rows and we are left with 2945 users. I assume that this is a sufficient number of entries for the model. Now the data is clean and ready for the next step.

### Transform data

In order to prepare data for the model we need to convert it to the proper format. The following code will convert non-numeric data to categories so that is it easier for the model to read it:

In [215]:
df['sex'] = df['sex'].astype('category').cat.codes
df['location'] = df['location'].astype('category').cat.codes
df['location_from'] = df['location_from'].astype('category').cat.codes
df['occupation'] = df['occupation'].astype('category').cat.codes
df['relationship_status'] = df['relationship_status'].astype('category').cat.codes

For the date of birth, I assume there is no need to keep the exact date - having just a year of birth should be enough for the model. I will drop the day and month information from 'dob' column:

In [216]:
df['dob'] = pd.DatetimeIndex(df['dob']).year

For the 'hobbies' column the best way is to extract dummies for each row and split it into columns with values of 0 or 1 indicating interest (or lack of interest) in a particular hobby. An additional 'hobby_' prefix will indicate that this column represents a hobby, but also to make sure that none of the column names are overlapping with the rest.

In [217]:
df = pd.concat([df.drop('hobbies', axis=1), df['hobbies'].str.get_dummies(sep=',').add_prefix('hobby_')], axis=1)

I am going to apply a similar procedure to the 'groups' column:

In [218]:
df = pd.concat([df.drop('groups', axis=1), df['groups'].str.get_dummies(sep='|').add_prefix('group_')], axis=1)

At this point the data contains only numbers, there are no missing values, and it is prepared for the next step.

In [219]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2945 entries, 1 to 3999
Columns: 3521 entries, user_id to group_muscle, strength  and vascularity - What every bodybuilder should know
dtypes: float64(1), int16(2), int64(3515), int8(3)
memory usage: 79.0 MB


user_id  ...  group_muscle, strength  and vascularity - What every bodybuilder should know
1        1  ...                                                  0                           
3        3  ...                                                  0                           
6        6  ...                                                  0                           
7        7  ...                                                  0                           
9        9  ...                                                  0                           

[5 rows x 3521 columns]

# Model Selection and Training

In this section I am going to build a model that will be used for predictions. For this task I am going to use Keras interface from the TensorFlow library, as well as some other libraries necessary for data validation and plotting.

In [220]:
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import confusion_matrix
import itertools
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

### Split data

For the model training and testing the dataset will be split into two subsets:
* 80% of the data will be used for training
* 20% od the remaining data will be used for testing

In [221]:
train_dataset = df.sample(frac=0.8, random_state=0)
test_dataset = df.drop(train_dataset.index)

I am going to separate labels from features:

In [222]:
train_features = train_dataset.copy()
test_features = test_dataset.copy()
train_labels = train_features.pop('target')
test_labels = test_features.pop('target')

### Build model

Now it is time to build a model. The selected model will be a regression-based neural network consisting of several input, hidden, and output layers. It will use existing data prepared in the previous section as an input in order to create predictions of the desired variable.

This model prefers to have the input data normalized in a specific way. Thus, we need to create a normalization layer that is adapted to the dataset:

In [223]:
normalizer = layers.experimental.preprocessing.Normalization()
normalizer.adapt(np.array(train_features))

Afterwards, we can build a fully-connected model consisting of a sequential stack of layers, where first layer is a normalization layer, then there are hidden layers using a rectified linear unit as an activation function, while the last output layer is using a sigmoid function in order to ensure our network output is between 0 and 1.

In [224]:
model = tf.keras.Sequential([
    normalizer,
    layers.Dense(3520, input_dim=3520, activation='relu'),
    layers.Dense(1, activation='sigmoid')
])
model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
normalization_2 (Normalizati (None, 3520)              7041      
_________________________________________________________________
dense_11 (Dense)             (None, 3520)              12393920  
_________________________________________________________________
dense_12 (Dense)             (None, 1)                 3521      
Total params: 12,404,482
Trainable params: 12,397,441
Non-trainable params: 7,041
_________________________________________________________________


### Compile and fit model



Next step is to compile the model. We need to specify the loss function to use for weights evaluation. As this is a binary classification problem we are going to use 'binary_crossentropy' as a loss function. We will classify the model using the collected accuracy value. We are also going to use Adam optimization algorithm for this model.

In [225]:
model.compile(
    optimizer=tf.optimizers.Adam(),
    loss='binary_crossentropy',
    metrics=['accuracy'])

Then, we can fit the model providing different settings that can be adjusted for the model efficacy.



In [226]:
history = model.fit(
    # Data to be used for training
    train_features, train_labels,
    # Number of epochs
    epochs=5,
    # Suppress logging
    verbose=0,
    # Calculate validation results on 20% of the training data
    validation_split = 0.2)

We can have a look at the last few epochs of the training of the model in order to see if everything works well.

In [227]:
# Show history in the last few epochs
hist = pd.DataFrame(history.history)
hist['epoch'] = history.epoch
hist.tail()

loss  accuracy  val_loss  val_accuracy  epoch
0  0.949993  0.773885  1.022934      0.832627      0
1  0.079435  0.968153  0.786424      0.858051      1
2  0.018477  0.995223  0.709390      0.860169      2
3  0.009324  0.998408  0.721772      0.858051      3
4  0.006124  0.999469  0.722202      0.860169      4

# Model Quality Assessment

In order to assess the quality of the model we are going to use the part of the dataset that hasn't been provided to the model yet.

In [174]:
test_predictions = model.predict(test_features)

Now we can see what is the fraction of correct prediction by comparing it to the true labels:

In [ ]:
correct = sum(i == j for i, j in zip(np.around(test_predictions), test_labels))[0]
print(correct / len(test_labels))

One can see that over 80% of predictions are correct. Even better way to look at the results is to create a confuction matrix showing the fraction of correct and incorrect predictions in each class (in this case it will be '0' and '1' as this is a binary classification).

In [ ]:
cm = confusion_matrix(test_labels, np.around(test_predictions))
cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
plt.imshow(cm, interpolation='nearest', cmap=plt.cm.Blues)
plt.title('Normalized confusion matrix')
plt.colorbar()
plt.ylabel('True label')
plt.xlabel('Predicted label')
thresh = cm.max() / 2.
for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
    plt.text(j, i, "{0:0.2f}".format(cm[i, j]),
        horizontalalignment="center",
        color="white" if cm[i, j] > thresh else "black")

One can see that over 90% of labels marked as '0' are correctly identified. For labels marked as '1' the model doesn't work that well and predicts correctly only around 28% of all cases.

# Scoring Test File

In this section the model will be used to produce target variable on the data stored in test.csv and test.json files. First, we need to read the file and transform it for the model in a similar way as before.

In [ ]:
df_model = df.copy()
df_csv = pd.read_csv(path+'test.csv')
df_json = pd.read_json(path+'test.json', orient='split').set_index('id')
df_groups = pd.DataFrame(columns=['groups'])
for i in df_json.to_dict()['groups'].items():
  groups = ''
  for j in i[1]['data']:
    if len(groups) > 0:
      groups += '|'
    groups += j['group_name']
  df_groups = df_groups.append({'groups': groups}, ignore_index=True)
df_groups.head()
df = pd.concat([df_csv, df_groups], axis=1)
df = df.drop(columns=['target', 'name', 'location_population', 'location_from_population', 'daily_commute', 'credit_card_type'])
df['hobbies'] = df['hobbies'].fillna('')
df['education'] = df['education'].fillna(df['education'].median())
df = df.dropna()
df['sex'] = df['sex'].astype('category').cat.codes
df['location'] = df['location'].astype('category').cat.codes
df['location_from'] = df['location_from'].astype('category').cat.codes
df['occupation'] = df['occupation'].astype('category').cat.codes
df['relationship_status'] = df['relationship_status'].astype('category').cat.codes
df['dob'] = pd.DatetimeIndex(df['dob']).year
df = pd.concat([df.drop('hobbies', axis=1), df['hobbies'].str.get_dummies(sep=',').add_prefix('hobby_')], axis=1)
df = pd.concat([df.drop('groups', axis=1), df['groups'].str.get_dummies(sep='|').add_prefix('group_')], axis=1)
df.info()
df.head()

Now we need to add columns that are missing from this dataframe and remove additional columns that didn't exist previously in order to have exactly the same set of columns as in the dataframe used for the model building. New columns will be filled with '0'. Additional column will be dropped because the model doesn't know what to do with hobbies or groups that didn't exist in the fitting data.

In [ ]:
missing_col = df_model.columns.difference(df.columns)
for col in missing_col:
  df[col] = 0
additional_col = df.columns.difference(df_model.columns)
df.drop(labels=additional_col.tolist(), axis=1, inplace=True)
df.drop(labels='target', axis=1, inplace=True)
df.info()

Now we can use the model to predict target value for the test dataset. As a result we get a list of probabilities:

In [ ]:
test_features = df.copy()
test_predictions = model.predict(test_features)
print(test_predictions)

We can finally prepare a scored test file. For the rows that we skipped in the prediction phase due to missing information we are going to assume the target variable as '0'. For the rest of the users we will use predicted probability and set a target value by rounding probability to the nearest integer value. Then, we save the output file as a CSV in a desired format consisting of 3 columns: 'user_id', 'probability_of_one', and 'target'.

In [ ]:
df_score = pd.DataFrame(columns=['user_id', 'probability_of_one', 'target'])
user_pred = 0
for user in range(df_csv.shape[0]):
  if user in df['user_id']:
    df_score =df_score.append({'user_id': user, 'probability_of_one': test_predictions[user_pred][0], 'target': np.around(test_predictions[user_pred][0])}, ignore_index=True)
    user_pred += 1
  else:
    df_score =df_score.append({'user_id': user, 'probability_of_one': 0., 'target': 0.}, ignore_index=True)
df_score = df_score.astype({'user_id': int, 'target': int})
df_score.head()
df_score.to_csv('test.csv', index=False)

# Findings

# Limitations of the Approach